Audio Improvements

Squareys opened the pull request:
Hi @mosra !

As usual, here is a work-in-progress pullrequest.

This pullrequest consist of:
 - Getters for `Source` and `Listener`
 - Tests of getters and setters of `Source` and `Listener`
 - Debug operator for `Source::DistanceModel`

**Moved to other branches:**
- [x] Add support for OpenAL extensions
   - [ ] Finish doc
   - [ ] Ensure Format Type parameter is actually a useful config setting.
   - [ ] Tests
- [x] Implement `AL_EXT_float32` and `AL_EXT_double`
- [x] Add support for the OpenAL soft hrtfs
- [ ] Implement OggAudioImporter (separate PR).

Greetings, Squareys

mosra commented on src/Magnum/Audio/CMakeLists.txt:
```diff
@@ -45,6 +45,11 @@ set(MagnumAudio_HEADERS
 
     visibility.h)
 
+if(WITH_SCENEGRAPH)
```
This is great. Exactly what I wanted to note but then I saw the conditional compilation is already done :)

Squareys replied:
Yeah, I figured you would obviously want to have Audio without SceneGraph sometimes.

mosra replied:
Yup, I wanted to fix that for DebugTools, if you enable them, they bring everything else and a kitchen sink now :)

mosra commented:
That's amazing amount of work you are doing here.

Re the naming: 

 - the Drawable feature was named after "what it is able to do" (drawable) or "what does it do" (draw), but the Audible is "what does it feel like", so it's a little inconsistent (with this approach the Drawable would have to be named Seeable). My first idea was Playable, more consistent with the rest, although it might be confused with Animable (but that should be okay, as it is in Audio namespace anyway)
 - the Listening feature - just Listener (similarly to how Camera is named)

OggAudioImporter - will you base it on stb_vorbis or the "official" libvorbis? I think it will be better to have it in magnum-plugins repo (keeping only the most essential plugins in the core repo).

Squareys commented:
> will you base it on stb_vorbis or the "official" libvorbis?

I would have used stb_vorbis as you suggested

mosra commented:
Sure! :)

Just a quick note about the `stb_*` libraries -- the original author is developing them as quick-and-dirty solutions to get up and running in demo and prototype code, they are small, easy to use, without dependencies, but they have some limitations and *might* not be as fast as the "official" decoders. That's no problem here, though -- for initial support this is good enough, very easy to implement, and if anyone would need the official "libvorbis" plugin, we can add it later.

The naming -- to be consistent with the rest, let's name the plugin `StbVorbisAudioImporter`, the class will be then `Audio::StbVorbisImporter` (similarly to the distinction in `WavAudioImporter`).

Squareys commented:
@mosra regarding the velocity in `Playable`. This would require access to a `Timeline` or something, correct? I could have flags for `position`, `velocity` and `orientation`, so that the `veloctiy` flag, for example, is only activated when a timeline has been set.

Squareys commented:
I added some commits for initial OpenAL extensions support. Still very sketchy and experimental and untested. But hrtfs work. 

I also added a Configuration struct for the Audio Context, that is very sketchy aswell. I might move Buffer::Format out of Buffer in case Configuration will *actually* need it. In that case I will take the necessary deprecation measures.

Squareys commented:
Okay, I rebased to have only the audio-improvement commits in this PR. Tests run and all code unrelated to improvements of existing code should have been moved away.

As soon as this is merged, I will properly rebase the other branches `audio-features` and `audio-extension-support` in my fork and file the work-in-progress pullrequests.

Good Night! :)

mosra commented on src/Magnum/Audio/Test/RendererTest.cpp:
```diff
@@ -34,10 +36,28 @@ struct RendererTest: TestSuite::Tester {
 
-RendererTest::RendererTest() {
-    addTests({&RendererTest::debugError});
+RendererTest::RendererTest(): _context{} {
```
No need to call `_context{}` explicitly (it will be done implicitly by the compiler).

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Test/SourceTest.cpp:
```diff
@@ -46,6 +71,102 @@ void SourceTest::debugState() {
 }
 
+void SourceTest::position() {
+    Source source{};
```
The {} at the end is superfluous, similarly in all other cases below.

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Test/RendererTest.cpp:
```diff
@@ -46,6 +66,63 @@ void RendererTest::debugError() {
+}
+
+void RendererTest::listenerGain() {
+    constexpr Float gain = 0.512;
```
Please use `f` suffix on floats (MSVC otherwise complains *a lot* and sometimes even refuses to compile that code).

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Test/RendererTest.cpp:
```diff
@@ -46,6 +66,63 @@ void RendererTest::debugError() {
+}
+
+void RendererTest::listenerOrientation() {
+    constexpr Vector3 up{1, 2, 3}, fwd{3, 2, 1};
```
Please use `{1.0f, 2.0f, 3.0f}` instead of integers. In combination with the {}-style initialization it causes implicit conversion warnings/errors on some compilers.

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Test/SourceTest.cpp:
```diff
@@ -46,6 +71,102 @@ void SourceTest::debugState() {
+void SourceTest::relative() {
+    Source source{};
+    source.setRelative(true);
+    CORRADE_COMPARE(true, source.isRelative());
```
```cpp
source.setRelative(true);
CORRADE_VERIFY(source.isRelative());
source.setRelative(false);
CORRADE_VERIFY(!source.isRelative());
```

In case of boolean values it's shorter to write :)

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Test/SourceTest.cpp:
```diff
@@ -46,6 +71,102 @@ void SourceTest::debugState() {
+
+void SourceTest::coneAnglesAndGain() {
+    Source source{};
+    constexpr Deg outerAngle = Deg(12);
```
```cpp
constexpr auto outerAngle = 12.0_degf;
```

Less typing :)

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Renderer.h:
```diff
@@ -188,6 +240,21 @@ class Renderer {
+/** @debugoperatorclassenum{Magnum::Audio::Renderer,Magnum::Audio::Renderer::DistanceModel} */
+Debug MAGNUM_AUDIO_EXPORT operator<<(Debug debug, Renderer::DistanceModel value);
+
+inline Vector3 Renderer::listenerPosition() {
```
I think you can put these directly into the class definition, they aren't too long to be hidden below.

Squareys replied:
- [x] *for keeping track*

mosra commented on src/Magnum/Audio/Renderer.h:
```diff
@@ -188,6 +240,21 @@ class Renderer {
+}
+
+inline std::array<Vector3, 2> Renderer::listenerOrientation() {
+    std::array<Vector3, 2> data = {Vector3{}, Vector3{}};
```
Initializing the array is not needed, it's done implicitly and will be overwritten by the getter below anyway.

Squareys replied:
- [x] *for keeping track*

mosra commented:
Commented on a few minor things inline, other than that I have no objections, great stuff :)

Please rebase on the latest master, the AppVeyor CI now builds also the Audio library and thus will be able to test this PR on MSVC 2015 right away.

Squareys commented:
Thanks! I will hopefully get around to fixing those by tomorrow.

Squareys commented:
Alright, the tests still fail on AppVeyor. I believe this must be a problem with the audio context (and device) not beeing created:

> Audio::Context: cannot open sound device OpenAL Soft
AL lib: (EE) ALCmmdevPlayback_open: Device init failed: 0x80070490

The things you pointed out have been fixed :)

Squareys commented:
I think `null` device could be the solution: https://github.com/kcat/openal-soft/blob/master/Alc/ALc.c#L104
That should be passed as a context configuration parameter, though. Which means that the `audio-extension` pullrequest needs to be finished first.

mosra commented:
Flawless! :)

The AppVeyor config is in [package/ci/appveyor.yml](https://github.com/mosra/magnum/blob/master/package/ci/appveyor.yml), commiting any change to it will immediately affect the AppVeyor build. The file has to be in `%AppData%`, right? So I think adding something like the following after [the block where OpenAL is extracted](https://github.com/mosra/magnum/blob/03c8272aea9ff5d30131786d906427bd81c00f9b/package/ci/appveyor.yml#L19) should do the trick:

```cmd
echo "[General]^r^ndrivers=null" > %AppData%/alsoft.ini
```

Guessing from the top of my head, might need to be slightly different, can't test right now.

**EDIT:** ugh, it seems like echoing a newline needs quite a lot of weird trickery: http://stackoverflow.com/questions/132799/how-can-you-echo-a-newline-in-batch-files Why, Windows, why? :D

Squareys commented:
Alright, awesome, will do so later today!

Squareys commented:
> (echo [General] && echo drivers=null) > %AppData%/alsoft.ini

Works for me locally, but AppVeyor does not like it. Maybe insufficient permissions?

LB-- commented:
Shouldn't that be `(echo [General] && echo.drivers=null)`? The period immediately after `echo` adds a newline.

Squareys commented:
Well... it works on my system, output is (and should be):
```
[General] 
drivers=null
```

Squareys commented:
Alright, works now. This is ready for merge then, once again.

mosra commented:
Merged, thank you a lot! :)
